master/internal/trial.go

```diff
@@ -363,58 +371,107 @@ func (t *trial) runningReceive(ctx *actor.Context) error {
 			return err
 		}
 
-	case actor.ChildFailed:
-		ctx.Tell(t.rp, scheduler.TerminateTask{TaskID: t.task.ID, Forcible: true})
+	case actor.ChildFailed, killTrial:
+		t.terminate(ctx, true)
+
+	case allReadyTimeout:
+		if time.Now().After(t.lastContainerConnectedTime.Add(allReadyTimeoutPeriod)) {
+			ctx.Tell(
+				t.logger,
+				model.TrialLog{
+					TrialID: t.id,
+					Message: "waited too long for all containers to be ready; " +
+						"this might happen when the k8s cluster waits for new pods to spin up, " +
+						"which might not happen at al; in the mean time, the resources that " +
+						"are already taken are not being used so you might want to cancel/kill " +
+						"this trial and restart it when the resources are available\n",
+				},
+			)
+		}
+
+	case terminateTimeout:
+		if msg.runID == t.runID {
+			ctx.Log().Info("forcibly terminating unresponsive trial after timeout expired")
+			t.terminate(ctx, true)
+		}
+
+	case actor.ChildStopped:
 
+	default:
+		return actor.ErrUnexpectedMessage(ctx)
+	}
+
+	return nil
+}
+
+func (t *trial) processSchedulerMsg(ctx *actor.Context) error {
+	switch msg := ctx.Message().(type) {
 	case scheduler.TaskAssigned:
 		if err := t.processAssigned(ctx, msg); err != nil {
 			return err
 		}
 
 	case scheduler.TerminateRequest:
-		ctx.Log().Info("trial runner requested to terminate")
-		t.pendingGracefulTermination = true
-
-	case sproto.ContainerStateChanged:
-		if msg.Container.State != container.Assigned {
-			t.startedContainers[msg.Container.ID] = true
-		}
-		switch msg.Container.State {
-		case container.Terminated:
-			t.processContainerTerminated(ctx, msg)
-		}
+		t.terminate(ctx, false)
 
 	case scheduler.TaskAborted:
 
 	case scheduler.TaskTerminated:
 		t.processTaskTerminated(ctx, msg)
 
-	case killTrial:
-		t.killTrial(ctx)
+	case scheduler.ContainerStarted:
+		t.containers[msg.Container.ID()] = msg.Container
 
-	case terminateTimeout:
-		if t.task != nil && msg.runID == t.runID {
-			ctx.Log().Info("killing unresponsive trial after timeout")
-			ctx.Tell(t.rp, scheduler.TerminateTask{TaskID: t.task.ID, Forcible: true})
+		if t.allReady(ctx) {
```

Comment:
non-blocking: i prefer the implementation where we call `allReady` within `pushRendezvous()`